src/sentry/sentry_metrics/indexer/cloudspanner/cloudspanner.py

```diff
@@ -60,28 +109,283 @@ def validate(self) -> None:
                 # TODO: What is the correct way to handle connection errors?
                 pass
 
+    def _get_db_records(self, db_keys: KeyCollection) -> Sequence[KeyResult]:
+        spanner_keyset = []
+        for organization_id, string in db_keys.as_tuples():
+            spanner_keyset.append([organization_id, string])
+
+        with self.database.snapshot() as snapshot:
+            results = snapshot.read(
+                table=self._table_name,
+                columns=["organization_id", "string", "id"],
+                keyset=spanner.KeySet(keys=spanner_keyset),
+                index=self._unique_organization_string_index,
+            )
+
+        results_list = list(results)
+        return [
+            KeyResult(org_id=row[0], string=row[1], id=self.__codec.decode(row[2]))
+            for row in results_list
+        ]
+
+    def _create_db_records(self, db_keys: KeyCollection) -> Sequence[SpannerIndexerModel]:
+        """
+        Create the set of database records from the provided dk_keys collection.
+        """
+        rows_to_insert = []
+        now = datetime.now()
+        for organization_id, string in db_keys.as_tuples():
+            new_id = get_id()
+            model = SpannerIndexerModel(
+                id=self.__codec.encode(new_id),
+                decoded_id=new_id,
+                string=string,
+                organization_id=organization_id,
+                date_added=now,
+                last_seen=now,
+                retention_days=_DEFAULT_RETENTION_DAYS,
+            )
+            rows_to_insert.append(model)
+
+        return rows_to_insert
+
+    def _insert_db_records(
+        self, rows_to_insert: Sequence[SpannerIndexerModel], key_results: KeyResults
+    ) -> None:
+        """
+        Insert a bunch of db_keys records into the database. When there is a
+        success of the insert, we update the key_results with the records
+        that were inserted. This is different from the postgres implementation
+        because postgres implementation uses ignore_conflicts=True. With spanner,
+        there is no such option. Once a write is successful, we can assume that
+        there were no conflicts and can avoid performing the additional lookup
+        to check whether DB records match with what we tried to insert.
+        """
+        try:
+            self._insert_batched_records(rows_to_insert, key_results)
+        except CloudSpannerRowAlreadyExists:
+            self._insert_individual_records(rows_to_insert, key_results)
```

Comment:
It's not very clear to me what we are doing with the rows that could not be inserted. Presumably if it fails a second time here it is because the orgid/string pair has already been written to the DB. Are we just ignoring it then?

Reply:
There should never be a situation where some row could not be inserted. One of the 2 api's should perform the insert. If it fails, the indexer should raise the exception(like network connectivity problems etc.)

Reply:
The scenario I'm referring to is when the org ID / indexed string unique constraint is not met. It seems plausible to me that two replicas try to insert it at the same time. Why is that not possible? Or is it being handled somewhere else in the code?

Reply:
istm what happens here is, we first try to batch insert, if that fails we insert records one by one so that we can handle conflicts better

Reply:
The logic now is as follows:
1. Perform a blind insert with all records. This can fail when a collision happens when 2 consumers try to write the same (org_id, string) or we are inserting with an existing id. If it succeeds, update the keyresult and move on. If it fails proceed to step 2.
2. Perform smart inserts. In a transaction context, get existing records for (org_id, string) tuple and existing records for (id) field. Make a new list of records to insert which don't have the records with existing records. Keep doing this in a loop until there's success.